### PR TITLE
senpi-helpers: fix restart for daemons launched via "python3 script.py"

### PR DESCRIPTION
**Pull request by rachinkapoor**
## Summary

Five batches of fixes shipped on this branch, derived from production failures observed today on **vulture** and **scorpion** and from in-depth code review of the senpi-helpers CLI.

| Batch | Commit | Theme |
|---|---|---|
| 1 | `5cd3a4f` | argv normalization — restart for `python3 -u script.py` daemons |
| 2 | `743ceb9` | log_path persists in boot.json + `start` command + pid-recycle guard + schema docs |
| 3 | `e2b11a9` | drop auto-resolve for destructive commands |
| 4 | `5cc63eb` + `400f2b2` | `boot` + `logs` inspection commands |
| 5 | `5793a46` | `diagnose` — composite pre-flight checklist |
| 6 | `7d4ab99` | `start --inherit-env-from <pid\|openclaw>` |

### Batch 1 — argv normalization (`5cd3a4f`)

`senpi-helpers restart` was failing silently on daemons launched the operator-documented way (`nohup python3 -u script.py &`). Two compounding bugs:

1. `state.write_boot` captured only `sys.argv`, missing the interpreter.
2. `manage.relaunch_daemon` did `Popen(argv)` which `execve`'s the .py file directly — fails with `PermissionError [Errno 13]` when the script isn't `+x`.

**Fix**: bump boot.json to schema 2 with `argv = [sys.executable, "-u", *sys.argv]`. Migrate schema-1 boot.json on read via `manage._normalize_argv` — legacy files keep working.

### Batch 2 — log_path + start + recycle guard (`743ceb9`)

Found while restarting vulture+scorpion live today: `senpi-helpers restart` was failing EARLIER than the argv bug — with `"no log path recorded"` (pid.json gone after clean exit, log_path was only there). Plus four more fixes:

- **boot.json schema 2 also persists `log_path`** so restart works after clean stop.
- **New `senpi-helpers start <name>`** command (idempotent peer of stop/restart).
- **Pid-recycle guard**: pid.json schema 2 records `cmdline_fingerprint` + `start_time_jiffies`. Every command consults `_pid_alive_for_daemon` which cross-checks both before signaling.
- **`ensure_daemon_stderr_redirected`**: dups stdout/stderr to `/tmp/<name>.log` when the daemon's stderr is a pipe (openclaw exec path). Linux-gated.
- **Schema versioning protocol** documented; `_read_json` warns on unknown schemas.

### Batch 3 — explicit name for destructive commands (`e2b11a9`)

`senpi-helpers stop` (no args) on a single-daemon host was auto-resolving and silently killing whatever daemon was there. Footgun for: wrong-box SSH, multi-window terminal confusion, muscle memory.

**Fix**: split `_resolve_name_readonly` (auto-resolves, used by list/health/stats/boot/logs/diagnose) and `_resolve_name_explicit` (NEVER auto-resolves, used by stop/restart/start).

No `[y/N]` prompt — openclaw's `exec` tool has no usable stdin and would hang. Same UX for human + agent: type the name.

### Batch 4 — `boot` + `logs` inspection (`5cc63eb` + `400f2b2`)

Two new read-only commands:

- **`senpi-helpers boot <name>`** — pretty-prints boot.json (argv, script_path, cwd, log_path, env_snapshot with long values truncated). `--json` for machine consumers.
- **`senpi-helpers logs <name> [-n N] [-f]`** — tail/follow the daemon log; resolves log_path via the same chain as restart/stats (pid → boot → env → /tmp default). `--follow` uses `tail -F` semantics (reopen on rotation, seek-to-0 on truncation, Ctrl-C clean exit).

### Batch 5 — `diagnose <name>` (`5793a46`)

Composite reader over pid.json + boot.json + heartbeat.json + log file. Replaces the manual ssh+cat+pgrep ritual when triaging a misbehaving daemon. Runs ~10 pure-function checks (pass/warn/fail + suggestion per check):

```
[✓] boot_json_present         boot.json schema 2
[✓] script_path_exists        /data/scripts/x.py exists
[✗] pid_alive_and_matches     pid 1234 is alive but cmdline / start_time fingerprint
                              doesn't match — kernel recycled it to a stranger
       → Run `senpi-helpers stop <name>` — the recycle guard will clear …
[!] log_file_exists           log file not found at /tmp/x.log
[✓] heartbeat_fresh           last tick 12s ago
[✓] last_tick_status          last tick status: ok
```

RC=0 when all checks pass; RC=1 when any fails (machine-friendly for cron/alerts).

### Batch 6 — `start --inherit-env-from <pid|openclaw>` (`7d4ab99`)

When starting vulture/scorpion manually today I had to:
1. `pgrep -f '^openclaw$'`
2. `cat /proc/<pid>/environ | tr '\0' '\n'` then `export` each line
3. THEN `nohup python3 -u <script>`

Now: `senpi-helpers start <name> --inherit-env-from openclaw`. Reads /proc/<openclaw>/environ, merges with the CLI's current env (operator-set values still win), passes to Popen. Linux-only; spec validation runs before the platform gate so dev-host typos still get useful errors.

## Live validation today (pre-merge)

Vulture + scorpion producers were both dead on the fleet. Reproduced two failures live on the boxes — the log_path issue (batch 2) hit FIRST, before the argv bug (batch 1) would have. Manual nohup launches restored both. They're ticking now with new pids:

- vulture pid 66914, daemon_started 2026-05-14T16:27:07Z
- scorpion pid 65709, daemon_started 2026-05-14T16:28:24Z

## Test plan

- [x] **283 tests green** (was 202 before this branch). +81 new, +3 updated assertions, 12 skipped on macOS for Linux-only paths.
- [x] `RUN_SLOW_INTEGRATION=1` full suite passes — real-subprocess reproducer for the original argv bug end-to-end.
- [x] Coverage: argv normalization, log_path fallback chain, recycle guard (positive + negative + Linux/macOS asymmetry), boot.json reader (+truncation), logs tail (+default lines), diagnose (each failure mode + all-passing), `--inherit-env-from` (pid + openclaw alias + invalid spec + Linux gate), auto-resolve disabled for destructive cmds.

## Backward compatibility

- Schema-1 `boot.json`: restart's `_normalize_argv` prepends interpreter on read. New daemon writes schema-2 on startup; migration is one-shot per daemon.
- Schema-1 `pid.json`: `pid_alive_and_matches` degrades to plain `pid_alive` when fingerprints are None.
- Existing consumers that pass `<name>` explicitly: unaffected. Auto-resolve removal only affects single-daemon-host TUI calls without a name.

## Verification (post-merge operator action)

```bash
# On vulture:
senpi-helpers diagnose vulture-producer-5ad1a1e05a2d
# Expect: all checks pass (it's running via my manual launch).

senpi-helpers stop vulture-producer-5ad1a1e05a2d
senpi-helpers start vulture-producer-5ad1a1e05a2d --inherit-env-from openclaw
# Expect: schema-1 boot.json migration banner + daemon_started in log.

senpi-helpers logs vulture-producer-5ad1a1e05a2d -n 20
# Expect: last 20 log lines from /tmp/vulture-producer.log.

senpi-helpers boot vulture-producer-5ad1a1e05a2d
# Expect: pretty-printed boot.json showing schema 2 + the new log_path field.
```

If green: scorpion + every other fleet box gets the same one-shot fix.

## What's deliberately out of scope (follow-up PRs)

From the code-review plan (`plans/03-...md` + `plans/04-...md`):
- Extract `_perform_relaunch` shared helper (internal refactor)
- `--strict-pid` flag (opt-in stricter recycle check)
- CLI-side lock for `cmd_start` (defensive only; no observed race)
- CI-on-Linux check (operator/infra)
- Tail-log-on-confirmation-timeout (workflow polish)

## Planning artifacts (in this PR's investigation tree)

- `debugging-tasks/engine_failuer_issue/plans/01-senpi-helpers-restart-fix.md` — batch 1 plan
- `debugging-tasks/engine_failuer_issue/plans/02-senpi-helpers-systemic-fixes.md` — batch 2 plan
- `debugging-tasks/engine_failuer_issue/plans/03-senpi-helpers-cli-deep-review.md` — code review
- `debugging-tasks/engine_failuer_issue/plans/04-senpi-helpers-cli-followups.md` — batches 3-6 plan

---

> [!NOTE]
> **Medium Risk**
> Touches core daemon lifecycle management (start/stop/restart), state-file schemas, and PID/log handling; mistakes could prevent daemons from restarting or could interfere with signaling. Changes are well-covered by expanded unit tests but still affect production ops paths.
> 
> **Overview**
> Fixes daemon relaunch reliability by bumping `boot.json`/`pid.json` to schema 2, adding legacy argv normalization (script-only → interpreter-first) and persisting `log_path` so `restart`/`stats` work after clean stops.
> 
> Hardens safety and operability: adds a pid-recycle guard (fingerprinted liveness via `/proc`), redirects daemon stderr to `/tmp/<name>.log` when launched with piped stderr, and disables implicit name auto-resolution for destructive commands (`start`/`stop`/`restart`).
> 
> Expands the CLI with new inspection and triage commands (`start`, `boot`, `logs`, `diagnose`) including `start --inherit-env-from <pid|openclaw>` for `/proc`-based env inheritance, plus shared log-path resolution and extensive new tests for tailing, diagnostics, and relaunch edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b8ea1efda66461f29e760b24c7315e5808920b7b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: senpi-trading-runtime/references/senpi-helpers-cli.md [hunk 1]
```diff
@@ -192,6 +192,60 @@ Exit codes:
...
+
+Exit codes:
+
+| code | meaning                                                |
+|------|--------------------------------------------------------|
+| 0    | started or already_running                             |
+| 1    | start failed (script path missing, spawn failure)      |
+| 2    | no `boot.json` for the daemon — never launched         |
+
+**First-time launches** still require the manual playbook (the daemon has
```
Comment (garg-prashant) [inline senpi-trading-runtime/references/senpi-helpers-cli.md:225]: This should be tested with multiple providers and models. If the LLM adheres to these instructions for lower models is good evaluation to have before pushing this to prod.

File: senpi-trading-runtime/references/senpi-helpers-cli.md [hunk 2]
```diff
@@ -277,6 +332,38 @@ What `restart` does, in order:
...
+### boot.json schema migration
+
+`boot.json` has two schemas in the wild:
+
+| schema | `argv` shape                                | how to launch |
+|--------|---------------------------------------------|---------------|
+| 1 (legacy) | `["/path/script.py"]`                   | `nohup python3 -u /path/script.py &` |
+| 2 (current) | `[sys.executable, "-u", "/path/script.py"]` | Written by every daemon since this fix |
+
+Why schema 2 exists: schema-1 captured only `sys.argv`, which is the
```
Comment (garg-prashant) [inline senpi-trading-runtime/references/senpi-helpers-cli.md]: This file is getting very bulky, very high changes for it to start hallucinating and not adhere to instructions during long conversation threads with the agent.

File: senpi-trading-runtime/senpi_runtime_helpers/manage.py [hunk 3]
```diff
@@ -202,28 +251,44 @@ def relaunch_daemon(
...
+    # subprocess.Popen take effect. The historical default-arg form bound
+    # subprocess.Popen at import time and silently bypassed test mocks.
+    if popen_factory is None:
+        popen_factory = subprocess.Popen
+
     if not argv:
         return {"outcome": RELAUNCH_SPAWN_FAILED, "pid": None,
-                "error": "argv is empty"}
+                "error": "argv is empty",
+                "argv_normalized": False, "argv_used": []}
```
Comment (garg-prashant) [inline senpi-trading-runtime/senpi_runtime_helpers/manage.py:278]: relaunch_daemon checks os.path.exists(argv[0]) before _normalize_argv and uses the current process CWD, not cwd from boot.json. If boot.json ever records a relative argv[0] (e.g. ./producer.py), this check can fail even though Popen(..., cwd=boot_cwd) would succeed, yielding a false RELAUNCH_SCRIPT_MISSING. Mitigation: resolve argv[0] against cwd (or only run exists after os.path.join(cwd or os.getcwd(), argv[0]) when the path is not absolute).

File: senpi-trading-runtime/senpi_runtime_helpers/cli.py [hunk 4]
```diff
@@ -552,8 +1077,279 @@ def _wait_for_new_pid_json(
...
+    Returns `(pid, None)` on success or `(None, error_message)` on failure.
+    Linux is the only supported target — `/proc` lookup happens via
+    `state.read_proc_environ`. The check for non-Linux happens at the
+    use-site so this helper stays pure.
+    """
+    if spec == "openclaw":
+        import subprocess
+        try:
+            res = subprocess.run(
+                ["pgrep", "-f", r"^openclaw$"],
```
Comment (garg-prashant) [inline senpi-trading-runtime/senpi_runtime_helpers/cli.py]: --inherit-env-from openclaw runs pgrep -f '^openclaw$', which matches only if the entire command line is exactly openclaw. Typical installs use paths (/usr/local/bin/openclaw) or subcommands (openclaw gateway, node …/openclaw). In those common cases pgrep returns nothing and the feature appears broken. Consider a pattern aligned with real cmdlines (e.g. basename or openclaw as a word), document the exact expectation, or match the parent process you actually need.

Comment (rachinkapoor) [reply senpi-trading-runtime/senpi_runtime_helpers/cli.py]: Fixed.

File: senpi-trading-runtime/references/senpi-helpers-cli.md (hunk 2 above)
Comment (rachinkapoor) [reply senpi-trading-runtime/references/senpi-helpers-cli.md]: Fixed

File: senpi-trading-runtime/references/senpi-helpers-cli.md (hunk 1 above)
Comment (rachinkapoor) [reply senpi-trading-runtime/references/senpi-helpers-cli.md:225]: Not sure what you mean here. 
Without a Producer there will no boot.json.
So what would start and restart do? 
Demon is producer specific.

File: senpi-trading-runtime/senpi_runtime_helpers/manage.py (hunk 3 above)
Comment (rachinkapoor) [reply senpi-trading-runtime/senpi_runtime_helpers/manage.py:278]: Fixed. Current code (lines 285–326 in `manage.py`):

1. `_normalize_argv(argv)` runs **first** — schema-1 `["script.py"]` → `[interpreter, "-u", "script.py"]`.
2. argv[0] is then resolved by category:
   - absolute path → check directly
   - relative-with-separator → `os.path.join(cwd or os.getcwd(), argv[0])` against **boot.json's cwd** (the same `cwd` Popen will use)
   - bare name → `shutil.which` against `$PATH`
3. `script_path` (when caller knows it) is also resolved against boot.json cwd as defense-in-depth — catches missing scripts in schema-2 argv where argv[0] is the interpreter, not the script.

Fix landed across:
- `e0fd059` — initial: resolve relative argv[0] against boot.json cwd
- `d2d73f6` — hardening: `os.path.isfile` (rejects empty / dir argv[0])
- `621f36c` — Bugbot follow-up: normalize **before** the existence check, so the check operates on Popen's actual argv (was rejecting bare-name `.py` argv via `shutil.which` before this).

File: senpi-trading-runtime/senpi_runtime_helpers/manage.py [hunk 5]
```diff
@@ -197,33 +248,108 @@ def relaunch_daemon(
...
+    elif os.sep in argv0 or (os.altsep and os.altsep in argv0):
+        # Has a path separator → operator meant a relative path. Join with cwd.
+        argv0_resolved = os.path.normpath(os.path.join(cwd or os.getcwd(), argv0))
+    else:
+        # No path separator → bare name. Honor $PATH the way Popen does.
+        which_result = shutil.which(argv0)
+        if which_result is None:
+            return {"outcome": RELAUNCH_SCRIPT_MISSING, "pid": None,
+                    "error": f"argv[0] {argv0!r} not found on $PATH",
+                    "argv_normalized": False, "argv_used": list(argv)}
```
[the remark this answers is not in the document]
Comment (rachinkapoor) [reply senpi-trading-runtime/senpi_runtime_helpers/manage.py]: This is the previous-review issue Bugbot is still tracking — it's already fixed.

The fix landed in `621f36c`: `_normalize_argv` now runs **before** the existence check. For schema-1 boot.json with `argv=["script.py"]`, the script gets prepended with `[interpreter, "-u"]` first, so the existence check operates on the interpreter (absolute path → `os.path.isfile` succeeds), not on a bare `.py` name routed through `shutil.which`.

See current `manage.py:285-326` and the reply on the parallel thread (#discussion_r3245139691) for the full resolved chain (e0fd059 + d2d73f6 + 621f36c).

File: senpi-trading-runtime/references/senpi-helpers-cli.md (hunk 1 above)
Comment (garg-prashant) [reply senpi-trading-runtime/references/senpi-helpers-cli.md:225]: What I meant was that a lot of instructions are in text format. For example, Launch a daemon as a detached process using argv + cwd recorded in `boot.json`. What if daemon is not launched via argv + cwd. Does that make a difference ?

The models don't have unlimited context and not every instruction it will pay attention to when the conversation threads get long. The textual context should only be used when there is a ambiguity situation and we need a decision making capability and interpretation. Could this be converted to a script ? 

Check for the existence of the file `boot.json` or start the daemon  ?

In case you still want to leave it on instructions level, then testing it with multiple models is important and that was the initial feedback.

File: senpi-trading-runtime/senpi_runtime_helpers/tests/test_cli.py [hunk 6]
```diff
@@ -415,8 +427,46 @@ def test_stats_missing_log_path_returns_no_log_exit(self) -> None:
...
+        os.makedirs(d, exist_ok=True)
+        with open(os.path.join(d, "pid.json"), "w") as f:
+            json.dump({
+                "schema": 1, "name": "default-log", "pid": os.getpid(),
+                "start_time_iso": "2026-05-12T08:00:00.000Z",
+                "wallet": "0x" + "a" * 40, "scanner": "s",
+                "interval_seconds": 300.0, "tick_timeout": 60.0,
+                "log_path": None, "version": "0.1.0",
+            }, f)
+        log_path = "/tmp/default-log.log"
```
Comment (garg-prashant) [inline senpi-trading-runtime/senpi_runtime_helpers/tests/test_cli.py:451]: This log path is different from `d` ?

**Review by garg-prashant — APPROVED** [review]
